components/eam/src/physics/cam/ndrop.F90

```diff
@@ -1620,49 +1620,37 @@ subroutine maxsat(zeta,eta,nmode,smc,smax)
    !      2. Multiple aerosol types. J. Geophys. Res., 105, 6837-6844.
 
    integer,  intent(in)  :: nmode ! number of modes
-   real(r8), intent(in)  :: smc(nmode) ! critical supersaturation for number mode radius
-   real(r8), intent(in)  :: zeta(nmode)
-   real(r8), intent(in)  :: eta(nmode)
-   real(r8), intent(out) :: smax ! maximum supersaturation
+   real(r8), intent(in)  :: smc(nmode) ! critical supersaturation for number mode radius [fraction]
+   real(r8), intent(in)  :: zeta(nmode) ! [dimensionless]
+   real(r8), intent(in)  :: eta(nmode) ! [dimensionless]
+   real(r8), intent(out) :: smax ! maximum supersaturation [fraction]
    integer  :: m  ! mode index
-   real(r8) :: sum, g1, g2, g1sqrt, g2sqrt
+   real(r8) :: sum, g1, g2
+   logical  :: lweak  !  BJG 10/7 whether forcing is sufficiently weak or not
 
-   do m=1,nmode
-      if(zeta(m).gt.1.e5_r8*eta(m).or.smc(m)*smc(m).gt.1.e5_r8*eta(m))then
-         !            weak forcing. essentially none activated
-         smax=1.e-20_r8
-      else
-         !            significant activation of this mode. calc activation all modes.
-         go to 1
-      endif
-   enddo
 
-   return
-
-1  continue
-
-   sum=0
+   lweak = .true.
    do m=1,nmode
-      if(eta(m).gt.1.e-20_r8)then
-         g1=zeta(m)/eta(m)
-         g1sqrt=sqrt(g1)
-         !BSINGH - repeated "g1=g1sqrt*g1" is a bug. Following code fixes this bug.
-         !BSINGH - This flag is added to maintain b4b result with the default code.
-         if(.not. fix_g1_err_ndrop) then
-            g1=g1sqrt*g1
-         endif
-         !BSINGH -ENDS
-         g1=g1sqrt*g1
-         g2=smc(m)/sqrt(eta(m)+3._r8*zeta(m))
-         g2sqrt=sqrt(g2)
-         g2=g2sqrt*g2
-         sum=sum+(f1(m)*g1+f2(m)*g2)/(smc(m)*smc(m))
-      else
-         sum=1.e20_r8
+      if(zeta(m).le.1.e5_r8*eta(m).and.smc(m)*smc(m).le.1.e5_r8*eta(m))then
+         lweak = .false.
```

Comment:
I think the core logic actually is the same:  

![Code_snippet](https://user-images.githubusercontent.com/48535869/194962885-cf813c1c-3aa9-4f83-93ba-3f52a2d734e4.png)

In both, one loops through all of the modes initially, and the forcing is considered weak unless for at least one mode the strong forcing condition is true (the negation of the if statement test condition in the original code).  If the forcing is weak for all modes, nothing is done other than set smax to a small value.  lweak is not overwritten each iteration of nmode, it's only overwritten if it happens to be true but for one mode it becomes false.

Having said that, the 'exit' statement does cause one difference between my code and yours / original:  as soon as it is determined that the forcing is not weak, the first loop is terminated, since further iterations are unnecessary.  This seems desirable, so I can go ahead and implement your version of the logic.

Reply:
Yeah, you are right. `lweak` value if becomes false, stays false. I also agree that the `exit` logic will save us from doing some unnecessary looping. Would you please re-implement it using the `exit` logic?

Reply:
OK, this has been done.